minio/minio.go

```diff
@@ -0,0 +1,198 @@
+		// create bucket
+		err = storage.CreateBucket()
+		if err != nil {
+			panic(err)
+		}
+	}
+
+	return storage
+}
+
+// Get value by key
+func (s *Storage) Get(key string) ([]byte, error) {
+
+	if len(key) <= 0 {
+		return nil, errors.New("the key value is required")
+	}
+
+	// check bucket
+	err := s.CheckBucket()
+	if err != nil {
+		return nil, err
+	}
+
+	// get object
+	object, err := s.minio.GetObject(s.ctx, s.cfg.Bucket, key, s.cfg.GetObjectOptions)
+	if err != nil {
+		return nil, err
+	}
+
+	// convert to byte
+	buf := new(bytes.Buffer)
+	_, err = buf.ReadFrom(object)
+	if err != nil {
+		return nil, err
+	}
+
+	return buf.Bytes(), nil
+}
+
+// Set key with value
+func (s *Storage) Set(key string, val []byte, exp time.Duration) error {
+
+	if len(key) <= 0 {
+		return errors.New("the key value is required")
+	}
+
+	// check bucket
+	err := s.CheckBucket()
+	if err != nil {
+		// create bucket
+		err = s.CreateBucket()
+		if err != nil {
+			return err
+		}
+	}
+
+	// create Reader
+	file := bytes.NewReader(val)
+
+	// set content type
+	s.cfg.PutObjectOptions.ContentType = http.DetectContentType(val)
+
+	// put object
+	_, err = s.minio.PutObject(s.ctx, s.cfg.Bucket, key, file, file.Size(), s.cfg.PutObjectOptions)
+
+	return err
+}
+
+// Delete entry by key
+func (s *Storage) Delete(key string) error {
+
+	if len(key) <= 0 {
+		return errors.New("the key value is required")
+	}
+
+	// check bucket
+	err := s.CheckBucket()
+	if err != nil {
+		return err
+	}
+
+	// remove
+	err = s.minio.RemoveObject(s.ctx, s.cfg.Bucket, key, s.cfg.RemoveObjectOptions)
+
+	return err
+}
+
+// Reset all entries, including unexpired
+func (s *Storage) Reset() error {
+
+	// check bucket
+	err := s.CheckBucket()
+	if err != nil {
+		return err
+	}
+
+	objectsCh := make(chan minio.ObjectInfo)
+
+	// Send object names that are needed to be removed to objectsCh
+	go func() {
+		defer close(objectsCh)
+		// List all objects from a bucket-name with a matching prefix.
+		for object := range s.minio.ListObjects(s.ctx, s.cfg.Bucket, s.cfg.ListObjectsOptions) {
+			if object.Err != nil {
+				log.Println(object.Err)
+			}
+			objectsCh <- object
+		}
+	}()
+
+	opts := minio.RemoveObjectsOptions{
+		GovernanceBypass: true,
+	}
+
+	for err := range s.minio.RemoveObjects(s.ctx, s.cfg.Bucket, objectsCh, opts) {
+		log.Println("Error detected during deletion: ", err)
+	}
+
+	return nil
+}
+
+// Close the storage
+func (s *Storage) Close() error {
+	return nil
+}
+
+// CheckBucket Check to see if already exist bucket
+func (s *Storage) CheckBucket() error {
+	exists, err := s.minio.BucketExists(s.ctx, s.cfg.Bucket)
+	if !exists || err != nil {
+		return errors.New("the specified bucket does not exist")
+	}
+	return nil
+}
+
+// CreateBucket Bucket not found so Make a new bucket
+func (s *Storage) CreateBucket() error {
+	return s.minio.MakeBucket(s.ctx, s.cfg.Bucket, minio.MakeBucketOptions{Region: s.cfg.Region})
+}
+
+// RemoveBucket Bucket remove if bucket is empty
```

Comment:
There's no way of deleting a bucket even if it has data?

Reply:
The bucket with data cannot be deleted, the data must be deleted first.